### ISSUE TITLE
current status

### ISSUE THREAD

**Issue by epcim** [CenturyLinkCloud/chef-provisioning-vsphere#2]
hi - I have tested it against my older vcenter and I got the following error. I use chefdk 0.6.x. How I may debug it? Are there any gem version requirements?

```
 Error executing action `converge` on resource 'machine[base-test-pemi]'
    ================================================================================
    
    RbVmomi::Fault
    --------------
    NotSupported: The operation is not supported on the object.
    

```

**Comment by mwrock** [comment]
Interesting. What version of vsphere are you running? We have run it on 5.0 and 5.5. One of the annoying aspects of the vsphere API is that some exceptions can be rather vague like this one. Do you mind sharing your recipe?

**Comment by epcim** [comment]
Well my script is a pretty simple same as a example. I only added the host section instead of resource_pool.

vsphere: 5.1.0 - build: 1882349

Notable to mention, I got some warnings as well:

```
[2015-05-26T20:32:20+02:00] INFO: Loading cookbooks []
Synchronizing Cookbooks:
Compiling Cookbooks...
Recipe: @recipe_files::/home/*********/hg2g/workspace-chef/vums/layouts/test-vsphere.rb
  * chef_gem[chef-provisioning-vsphere] action install[2015-05-26T20:32:20+02:00] INFO: Processing chef_gem[chef-provisioning-vsphere] action install (@recipe_files::/home/********/hg2g/workspace-chef/vums/layouts/test-vsphere.rb line 2)
WARN: Unresolved specs during Gem::Specification.reset:
      httpclient (>= 2.2.0.2, ~> 2.2)
      nokogiri (>= 1.4.0, ~> 1.5)
      akami (~> 1.0)
      gyoku (>= 0.4.0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
 (up to date)
[2015-05-26T20:32:23+02:00] WARN: Node nodeone has an empty run list.
  Converging 2 resources
  * chef_gem[chef-provisioning-vsphere] action install[2015-05-26T20:32:23+02:00] INFO: Processing chef_gem[chef-provisioning-vsphere] action install (@recipe_files::/home/*******/hg2g/workspace-chef/vums/layouts/test-vsphere.rb line 2)
WARN: Unresolved specs during Gem::Specification.reset:
      trollop (>= 0)
      mini_portile (~> 0.6.0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
 (up to date)
  * machine[base-test-pemi] action converge[2015-05-26T20:32:23+02:00] INFO: Processing machine[base-test-pemi] action converge (@recipe_files::/home/******/hg2g/workspace-chef/vums/layouts/test-vsphere.rb line 39)

```

```
chef_gem 'chef-provisioning-vsphere' do                
  action :install
  compile_time true
end

require 'chef/provisioning/vsphere_driver'

with_vsphere_driver host: 'vcenter.intranet.lab',
  insecure: true,
  user:     '*********',
  password: '**********'

#vagrant/ubuntu1204-mini',
with_machine_options :bootstrap_options => {
  use_linked_clone: true,
  num_cpus: 1,
  memory_mb: 1024,
  network_name: ["VM Intranet"],
  datacenter: 'Datacenter',
  #resource_pool: 'lab.ci',
  host: '172.16.0.43',
  vm_folder: 'vagrant-chef-ci',
  datastore: 'dsm3300r5sas1',
  template_name: 'small-ubuntu1204.lab.ci',
  customization_spec: {
    ipsettings: {
      dnsServerList: ['8.8.8.8']
    },
    :domain => 'local'
  },
  :ssh => {
    :user => 'root',
    :password => 'password',
    :paranoid => false,
  }
}

machine "base-test-pemi" do
  run_list ['lab_base::default']
  #recipe 'lab_base::default'
end                                               
```

**Comment by mwrock** [comment]
Could you provide the chef-stacktrace.out output?

**Comment by CBeerta** [comment]
Running into the same issue, on vSphere 5.1  (Build 1364037)

Same simple test script causes the Exception "The operation is not supported on the object." (In german though):

```
Generated at 2015-06-09 09:41:07 +0200
RbVmomi::Fault: machine[my_machine_name] (@recipe_files::/home/cbeerta/src/git/chef/mcs-chef/vsphere.rb line 32) had an error: RbVmomi::Fault: NotSupported: Der Vorgang wird für dieses Objekt nicht unterstützt.
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/vim/Task.rb:11:in `wait_for_completion'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/vsphere_helpers.rb:251:in `block in create_delta_disk'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/vsphere_helpers.rb:233:in `each'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/vsphere_helpers.rb:233:in `create_delta_disk'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/vsphere_helpers.rb:220:in `relocate_spec_for'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/vsphere_helpers.rb:139:in `do_vm_clone'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/driver.rb:494:in `clone_vm'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/driver.rb:177:in `find_or_create_vm'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/driver.rb:135:in `allocate_machine'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-1.2.0/lib/chef/provider/machine.rb:30:in `block in <class:Machine>'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-1.2.0/lib/chef/provider/machine.rb:37:in `block in <class:Machine>'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
/home/cbeerta/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-1.2.0/lib/chef/provider/machine.rb:55:in `block in <class:Machine>'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
/opt/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
/opt/chefdk/embedded/apps/chef/lib/chef/provider.rb:130:in `run_action'
/opt/chefdk/embedded/apps/chef/lib/chef/resource.rb:562:in `run_action'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:49:in `run_action'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block (2 levels) in converge'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `each'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block in converge'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:83:in `block in execute_each_resource'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:81:in `execute_each_resource'
/opt/chefdk/embedded/apps/chef/lib/chef/runner.rb:80:in `converge'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:339:in `block in converge'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `catch'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `converge'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:353:in `converge_and_save'
/opt/chefdk/embedded/apps/chef/lib/chef/client.rb:457:in `run'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:271:in `block in fork_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:259:in `fork'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:259:in `fork_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:225:in `block in run_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:402:in `block in interval_run_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `loop'
/opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `interval_run_chef_client'
/opt/chefdk/embedded/apps/chef/lib/chef/application/client.rb:382:in `run_application'
/opt/chefdk/embedded/apps/chef/lib/chef/application.rb:60:in `run'
/opt/chefdk/embedded/apps/chef/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:51:in `load'
/usr/bin/chef-client:51:in `<main>'
```

I've set use_linked_clone to false, but no difference.

**Comment by mwrock** [comment]
Thanks for the stack trace! This trace appears to be bombing out creating the delta disk for the linked clone. You mentioned that you see the same error if you set `use_linked_clone` to `false`. I'm assuming the above stack trace was generated on a run when `use_linked_clone` was set to `true`. Could you send another stack trace when it is set to `false`?

Also could you tell me if you are providing the name of an actual vm template or just another shut down vm as the `template_name`?

Thank you!

**Comment by CBeerta** [comment]
No, the trace was already with ```use_linked_clone``` set to false: 

```
    - creating machine my_machine_name on vsphere://vc-testlab.testlab.mcs.de/sdk?use_ssl=true&insecure=true
    -   use_linked_clone: false
    -   num_cpus: 2
    -   memory_mb: 512
    -   network_name: ["kein Uplink"]
    -   datacenter: "Testlab"
    -   vm_folder: "Chef-Test"
    -   host: "Testcluster"
    -   datastore: "datastore"
    -   template_name: "ubuntu-14"
    -   customization_spec: {}
    -   ssh: {:user=>"root", :password=>"secret", :paranoid=>```

The Template is an actual Template that was converted from a PXE installed Ubuntu VM.

**Comment by mwrock** [comment]
I now see a definite bug where we use a linked clone if use_linked_clone is in the settings at all true or false. You can get around this for now by simply not specifying the use_linked_clone setting at all.

I'm also guessing but need to confirm that linked clones will only work on a full fledged clone and not a template. If that is the case we need to provide better messaging/errors.

**Comment by CBeerta** [comment]
Yes, that indeed did it. Just removing the use_linked_clone line. I'm not able to use the  'host'  Option, only a resource pool, but that is already in #4.


Thanks

**Comment by mwrock** [comment]
just released 0.5.7.dev which should take care of the host issue you may be experiencing and also ensures that linked clones are not used when explicitly setting `use_linked_clone` to `false`.

I'm leaving this open to investigate using linked clones with templates.

**Comment by mwrock** [comment]
Just to follow up on this issue. I have determined that linked clones are not supported on VM Templates but now the driver catches if you are using a template and simply emits a warning that it is ignoring the linked template.

This is available in the latest dev gem 0.5.7.dev6 and will roll up in to 0.6.0 when it releases.

**Comment by EdHurtig** [comment]
I find it very interesting that linked cloning from a protected template is explicitly not supported, I would expect it to be the exact opposite as you would likely want to mark a virtual machine used for linked clones as a template to protect the root snapshot that all linked clones are based on.

@mwrock Where did you find that templates are explicitly unsupported on recent versions of vSphere?  I can confirm that the standard linked clone setup works fine on vSphere 6.0 as far as I can tell

**Comment by mwrock** [comment]
I wish I could remember exactly @EdHurtig. I'm pretty sure the problem occurs here: https://github.com/CenturyLinkCloud/chef-provisioning-vsphere/blob/master/lib/chef/provisioning/vsphere_driver/vsphere_helpers.rb#L132.

Creating the delta disk from a vm template's disk. I believe this was only successful against a full fledged parent VM. I don't have access to vsphere infrastructure these days to validate or experiment with.

**Comment by EdHurtig** [comment]
Huh, that is interesting because I did some digging and heard similar "must be a fully fledged VM" statements around, but I cannot reproduce any of those problems on 6.0, but that doesn't mean the problem doesn't exist for other vSphere setups.

I forked and reformatted so that it showed a warning, but does not actually prevent linked cloning from a template. Working great so far.  

Please let me know if you would like me to PR those changes back here, otherwise I am happy to live dangerously on a fork :D

**Comment by mwrock** [comment]
Interesting. I had only used with 5.0 and 5.5 so perhaps its an enhancement from 6.0.

**Comment by EdHurtig** [comment]
perhaps they did fix this with 6.0

**Comment by bhaliti** [comment]
Hi,

I am on chef-provisioning-vsphere-0.9.0 using vsphere 5.5 and have removed used linked clone but am still getting this error. tried with and without it. 

================================================================================
Error executing action `allocate` on resource 'machine[firstmachine]'
================================================================================

NoMethodError
-------------
undefined method `[]' for nil:NilClass

any help would be appreciated

thanks
